### ISSUE TITLE
Fix S3604 FP: Primary constructors

### ISSUE THREAD

**Issue by hankovich** [SonarSource/sonar-dotnet#7624]
### Description

S3604 is reported for classes with primary constructors.

### Repro steps

```csharp
class SampleClass(object options)
{
    private readonly object _options = options; // Noncompliant - FP: S3604 shouldn't be reported here, this is not a redundant initializer
}
```

### Expected behavior

S3064 should not be raised.

### Actual behavior

S3064 raises a warning.

### Known workarounds

N/A

### Related information

* C#/VB.NET Plugins version: 9.5.0.73987
* Visual Studio version: latest preview
* MSBuild / dotnet version: .NET 8 preview 6

**Comment by zsolt-kolbay-sonarsource** [comment]
Thank you for reporting this issue @hankovich. Confirmed as False Positive.

**Comment by smokedlinq** [comment]
With .NET 8.0 right around the corner, this would be nice to get fixed as the primary constructor refactor causes this to appear

**Comment by OpenEnergyGroup** [comment]
This is an issue for properties too: this code will raises an incorrect **S3604** warning:

```
public class Foo(string foo)
{
    public List<string> MyList { get; } = [foo];
}
```

Thanks.

**Comment by Guiorgy** [comment]
.NET 8 is out, and indeed I get `S3604`:

```cs
public sealed class MyClass(string name, string lastName) : IEquatable<MyClass>
{
    public string Name { get; } = name; // S3604
    public string LastName{ get; } = lastName; // S3604
    // ...
}
```

As a workaround, you could rewrite the properties to return the arguments dirrectly instead of the backing fields:

```cs
public sealed class MyClass(string name, string lastName) : IEquatable<MyClass>
{
    public string Name { get => name; }
    public string LastName{ get; } = lastName; // S3604
    // ...
}
```

`Name` no longer triggers `S3604`. Here's the Generated C#:

```cs
[System.Runtime.CompilerServices.NullableContext(1)]
[System.Runtime.CompilerServices.Nullable(0)]
public sealed class New
{
    [CompilerGenerated]
    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
    private string <name>P;

    [CompilerGenerated]
    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
    private readonly string <LastName>k__BackingField;

    public string Name
    {
        get
        {
            return <name>P;
        }
    }

    public string LastName
    {
        [CompilerGenerated]
        get
        {
            return <LastName>k__BackingField;
        }
    }

    public New(string name, string lastName)
    {
        <name>P = name;
        <LastName>k__BackingField = lastName;
        base..ctor();
    }
}
```

So pretty much the same.

**Comment by Corniel** [comment]
I can confirm this issue.

**Comment by Davidsv** [comment]
Still got this issue, NET 8.0.100, latest visal studio 17.8, latest sonarlint 7.4.0.80741

**Comment by FelixDamrau** [comment]
> .NET 8 is out, and indeed I get `S3604`:
> _...snip..._
> As a workaround, you could rewrite the properties to return the generated private field dirrectly instead of the backing fields:
> _...snip..._
> So pretty much the same.

The workaround suppresses the S3604 message, but the compiler-generated backing field is not `readonly` anymore.

Thank you for your effort in providing the workaround. However, I still need the `readonly` stuff, so this issue remains a concern for me.

_Anyhow, I mainly want to add a comment, so this issue may get prioritized._

**Comment by Guiorgy** [comment]
You're right, missed that. `readonly` should allow the runtime some low level optimizations too

**Comment by shkarface** [comment]
The issue has affected our codebases as we're upgrading to .NET 8 and new C# features as well.

+1.

**Comment by ms-lemos** [comment]
+ 1. False positives with .net 8. Commenting in hope for this to get priority, it's quite annoying.

**Comment by andrei-epure-sonarsource** [comment]
Thanks, we will prioritize it and apologies for missing this in our previous release.

**Comment by Corniel** [comment]
> Thanks, we will prioritize it and apologies for missing this in our previous release.

Those things can happen. No worries. Are you considering to ship it earlier than you normally do?

**Comment by andrei-epure-sonarsource** [comment]
@Corniel it went into the latest release. Soon we'll deploy to nuget.org and next week to SonarCloud.
This will be included in SQ 10.4 and the next release of SonarLint.

**Comment by Corniel** [comment]
We're all developers, we all miss things sometimes. No reason to apologize. Keep up the good work!

**Comment by gho1b** [comment]
This issue still happen in `v9.9.8` (lts-community) when i try the same logic like this one
```C#
public abstract class DbService<TDbContext>(IDbContextFactory<TDbContext> factory) : IDbService
    where TDbContext : DbContext
{
    private bool _disposed;
    private readonly Lazy<TDbContext> _db = new(factory.CreateDbContext);
```

When i try to implement the complian solution, it raise warning _Use primary constructor[IDE0290](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/style-rules/ide0290)_
```C#
public abstract class DbService<TDbContext> : IDbService
    where TDbContext : DbContext
{
    private bool _disposed;

    private readonly Lazy<TDbContext> _db;

    public TDbContext Db => _db.Value;

    protected DbService(IDbContextFactory<TDbContext> factory)
    {
        _db = new Lazy<TDbContext>(factory.CreateDbContext());
    }
```

Am i missing something here?

**Comment by sebastien-marichal** [comment]
Hello @gho1b,

The C# analyzer included in SonarQube Server v9.9.8 is version 8.51. This fix is available starting from version 9.14.

To obtain this fix, please update your SonarQube Server instance to version 10.4 or higher.
